Introduce new power management

kaspar030 opened the pull request:
I finally got around to code some ideas we've been discussing in the past for improving our power management.

Reasons:
- current LPM us completely unused, only idling is enabled, more sleep modes are hard commented out even if implemented.
- current LPM is actually completely unusable as there's only one global "wakelock" for every power mode

This is intentionally kept simple in order to make it implementable for a wide range of platforms. IMHO It will never be possible to find an API that supports all the possible power saving mechanisms on all platforms (enabling/disabling/changing clock sources for everything, sleepwalking), but it is possible to save a lot of power compared what we currently have, and I also believe it is possible to come close enough to minimal power usage to be usable.

From doxygen:

```
This simple power management interface is based on the following assumptions:

- CPUs define up to 4 power modes (from zero, the lowest power mode, to
  PM_NUM_MODES-1, the highest)
- there is an implicit extra idle mode (which has the number PM_NUM_MODES)
- individual power modes can be blocked/unblocked, e.g., by peripherals
- if a mode is blocked, so are implicitly all lower modes
- the idle thread automatically selects and sets the lowest unblocked mode

The following functions need to be implemented for each platform:

pm_init()
pm_set()
```

I've selected SAM R21 as test board to try an actual implementation.

*WARNING* the test application currently sets STANDBY mode at some point, from which openocd cannot flash the board anymore. I have to use edbg to clean the flash before retrying.

kaspar030 commented:
Open questions:

- @immesys I'm very interested in your changes you've mentioned in #2309 to get the power consumption down!

- how do we integrate turning off a board? Seems overkill to reserve wakelock memory even if that is usually just another "low power mode"

- while the PM module from this PR is currently completely optional, should we just remove the old lpm infrastructure at the same time? It is not used anywhere for more than idling the CPU, anyways.

- are four power modes (plus CPU idle, plus maybe OFF) enough?

kaspar030 commented:
The test application just tries xtimer_spin() and xtimer_sleep() for 5 seconds each, unblocking power modes every 20 secs. As soon as mode 0 is unblocked, after the spinning completes, the nodes basically shuts down, as expected.

My measurement equipment is rather crude (cheap multimeter with uCurrent gold, samr21 gets power with the current measurement header piped through the ucurrent).

Power consumption currently is as follows:

|Mode |uA | notes |
|--------|-----|---------|
| spinning | 19 | LED on | 
| idle (3) | 14.8 | |
| 2 | 13.4 | |
| 1 | 12.2 ||
| 0 | ~10.5 | fluctuates between 9.5 and 11.5, system stuck |

miri64 commented:
Not sure this is a level to low, but aometimes an action after wake-up is required (especially for network or MAC protocols). See e.g. [6Lo-ND's wakeup behavior](https://tools.ietf.org/html/rfc6775#section-5.8.2). Are there plans to integrate the pending event management into this?

kaspar030 commented:
> Not sure this is a level to low, but aometimes an action after wake-up is required (especially for network or MAC protocols). See e.g. 6Lo-ND's wakeup behavior. Are there plans to integrate the pending event management into this?

Hmm, that would require *every* ISR to call a hook. What are the exact requirements? Just "whenever a node wakes up from a low power mode, do something"? That seems very unspecified when thinking about the possible implementations of sleeping.

miri64 commented:
Well my idea was to just send out some IPC message (or have some generic callback like in `xtimer`). From the general to the specific: The exact requirements for 6Lo-ND's wakeup behavior is: iff the next wakeup happens after the registration timeout: re-new registration now by sending an NS with an ARO out to the default router.

miri64 commented:
(registration lifetime is in order of minutes, so this happens rather rarely per node)

lebrush commented:
A my grain of sand. I attended to a presentation by TI during electronica and they presented their concept for LP Embedded OS which I found very interesting. The rough summary follows:

 - Consider a MCU specific LP-Manager (detailed below)
 - On the implementation of every peripheral they register the required elements for their functionality to the LP-Manager (i.e. I need DMA, Timer, ADC and be able to wake-up on interrupt).
 - When the use of the peripheral is done, they communicate it to the LP-Manager, so the constraint is no longer considered.
 - During the idle-thread the LP-Manager is polled and the lowest possible LP is selected.
 - On "wake-up" the timers are adjusted (if required) to compensate the delay. 

Imagine a SPI is configured in master mode. When initialized we don't care, but when starting a communication, it will tell the LP-Manager, not to turn off the SPI clock and i.e. to keep the DMA and it's interrupts enabled. The LP-Manager will go to the lowest power mode possible within this constrains while the transfer is ongoing. When the SPI is done, the LP-Manager knows the constrains are no longer applicable and can send the MCU to an even lower LPM.

I'm not sure it's understandable. Let me know if I can improve the description or give more details :-)

kaspar030 commented:
> I'm not sure it's understandable. Let me know if I can improve the
> description or give more details :-)

I think it's understandable enough :)

This PR is currently missing any actual blocking/unblocking of power
modes, which I envision to be done by the periph implementations.

E.g., the RTC might need the low-power OSC, which on a hypothetical MCU
keeps running in power mode 0, while a high-speed timer needs the
high-speed OSC, which is only available in power mode 2. Both would now
need to block/unblock the corresponding power modes when a timer is
set/fired.

In a way, the actual MCU-specific implementation of this PR's interface,
combined with the configuration, is similar to the LP-Manager TI is
sketching.

The main difference is that this PR's design considers it's power modes
to be layered instead of basically being able to turn on/off each used
peripheral's clocks/resources individually.

I think this constraint makes implementing usable power saving far easier.

Once periphs actually do block/unblock calls as configured in e.g.,
periph_conf.h, it would also be simple to remove the layering
constraint, should it prove to increase power savings substantially.

kaspar030 commented:
> Well my idea was to just send out some IPC message (or have some generic callback like in xtimer).

I think this is unrelated to actual power management. We could, before going to sleep, disable thread_yield() somehow, then fire the wakeup hooks (send messages, set thread flags, unlock mutexes, whatever). That would put the threads in pending mode, but not trigger the actual context switch. On wakeup the threads would get scheduled right after the ISR's are handled.

lebrush commented:
@kaspar030 good point, it makes it way easier to implement :-)  

I understand that the plan is to have several LPM which are MCU dependent and not anymore RIOT generic, right? Two comments with the current approach:

 - each peripheral needs to know what peripherals are on/off in each power mode
 - if I only need an interrupt in the GPIO port A (i.e. stm32) this would keep the clock running in the other ports as well. In case of the stm32l1, disabling the clock and setting the pins floating of each port has a huge impact in the power consumption.

In any case I think is a really good approach and I think is the right way to go from the current standing point. Good job! :-) Maybe once this first step is implemented, we can evaluate how much work is (and if it is worth) to go for a individual peripheral selection (TI approach), which at the moment is far too ambitious.

kaspar030 commented:
> I understand that the plan is to have several LPM which are MCU
> dependent and not anymore RIOT generic, right?

Yup!

>   * each peripheral needs to know what peripherals are on/off in each
>     power mode

Yup, or, it needs to know which power mode it needs to block. That can
be configured in e.g., periph_conf.h.

>   * if I only need an interrupt in the GPIO port A (i.e. stm32) this
>     would keep the clock running in the other ports as well. In case of
>     the stm32l1, disabling the clock and setting the pins floating of
>     each port has a huge impact in the power consumption.

That would depend on the gpio implementation (and its configuration).
It should be possible to say

    #define GPIO_PORT_A_POWER_MODE (2)
    #define GPIO_PORT_B_POWER_MODE (4)

... and then block/unblock accordingly, automatically, maybe in
gpio_init_int().

> In any case I think is a really good approach and I think is the right
> way to go from the current standing point. Good job! :-)

Thanks! ;)

immesys commented:
I have extensive experience with getting the power consumption low on the SAMR21 and the SAM4L. Both with TinyOS and with RIOT. My previous platform had an idle current of 2.3uA on SAM4L and my present one has 2.6uA on SAMR21 (where idle is defined as in xtimer_usleep for example). As it turns out the experience is a little different between them. In the SAMR21/SAMD21 Atmel put a lot of effort into things like ONDEMAND bits for clocks. To go into low power mode you mostly just go into the lowest power mode every time and the peripherals themselves will take care of keeping the correct clocks active. For the SAM4L it was a royal pain to make sure peripherals had the correct clocks active and you entered the right depth of sleep. When you woke up, every ISR that might be the wakeup source had to call into a hook that would perform the wake-up-dance of re-enabling the correct clocks.

I will caution very strongly against the approach of "poll a bunch of things to determine the lowest sleep state in the idle thread". TinyOS did that because on msp430 (and other earlier platforms) the amount of state that required polling was small. On modern platforms, the number of peripherals and the complexity of the state they may be in (different clocks depending on how the peripherals are configured) makes it far more difficult to poll the state and you really don't want to spend more than a handful of cycles working out how to go to sleep because it wastes a lot of energy.

I have more than 100 motes deployed running RIOT. They have an average current of about 22uA sampling and sending every 10-30s. They ought to last about 5-10 years on the batteries they have. You can see the changes I had to make to RIOT to get that working here:

https://github.com/hamilton-mote/RIOT-OS/compare/b333b9bc27d0643fe7ed34e7a3f4c003034c5726...hamilton-mote:hamilton-lp-patches

The TLDR is:

- use OSCULP32K as it runs always during sleep anyway
- use the RTT to back xtimer, not a high-power-timer
- change the GPIO EIC clock to OSCULP32K not GCLK0 as that doesn't run during sleep so you won't wake up
- Change the app to put the radio to sleep after sending (we really need decent MACs in RIOT...)

immesys commented:
@kaspar030 I can send you a SAMR21 board that is better suited for power measurement if you want, or revive my remote-measurement jig. The SAMR21XPRO has a lot of other stuff of in that ruins measurements of just the MCU.

haukepetersen commented:
Nice, I very much like the new/reworked concept!

> how do we integrate turning off a board? 

How about we introduce a explicit function for this? While at it, how about also including the reboot function:
```c
void pm_reboot(void);
void pm_halt(void);
```

> while the PM module from this PR is currently completely optional, should we just remove the old lpm infrastructure at the same time? It is not used anywhere for more than idling the CPU, anyways.

Yes, I am 100% for dropping the old lpm stuff right away -> otherwise it will take years again until stuff converges.

> are four power modes (plus CPU idle, plus maybe OFF) enough?

Could well be, but I would tend to go into a slightly different direction: from what we have learned and know, the low power stuff is highly dependent on the CPU. So I think it is best to make the CPU define the actual structure of the  `pm_blocker_t` struct. Also I tend to think that it makes sense, have less (or no?) code except the interface in core and let the CPUs care about everything. So how about such an interface:
```c
void pm_block(unsigned mode);
void pm_unblock(unsigned mode);
void pm_set_lowest(void);
void pm_halt(void);
void pm_reboot(void);
```
The actual definition and structure of the blocker union could be completely internal to the actual CPU specific implementation, allowing for any number of power states. What do you think about that?

LudwigKnuepfer commented:
I think I like the concept in general.

However I think there needs to be some generic concept that enables things like xtimer (I can't think of anything else right now, but there might be others) to communicate the time until the next wakeup. The reason is that there generally is some overhead associated with switching power modes and clock frequencies, so the power manager needs to know if it is "reasonable" to switch at all.

immesys commented:
@LudwigKnuepfer My 2c, 99% of the time it is fine to have a policy of "immediately try to go to the lowest power mode allowed when you hit the idle thread". I wouldn't worry about the overhead of going to sleep when you are just about to wake up. It is a rare occurrence and you are likely to spend more energy on the cycles checking for that condition than you are going to save.

LudwigKnuepfer commented:
There is not only energy consumption but also timeliness involved.

haukepetersen commented:
@kaspar030: had any time to update the PR?

@LudwigKnuepfer: timeliness is indeed something that needs to be clearly handled. For the (current version) of the 'xtimer' it is quite simple: the xtimer uses low-level drivers, which block the system from going to deep-sleep anyway, so not timing problems here :-). But as this PR changes the paradigm from 'pm tells x y z to shut down because we want to sleep` to 'x y z tell the PM how deep it can go to sleep', the modules depending on some (CPU specific) delays (or not to have them), one should be able to configure the system in a way, so that all timing demands can be met. But this behavior needs to be actually verified in practice...

haukepetersen commented:
any news on 'periphying' this?

haukepetersen commented on core/include/pm.h:
```diff
@@ -0,0 +1,127 @@
+ *
+ * @param[in]   mode      power mode to block
+ */
+static inline void pm_block(unsigned mode)
```
just played with this -> IMHO it makes sense not to inline the (un)block functions to save some code size, and as they are usually not called in a time critical path?!

kaspar030 commented:
Please help getting this merged, so we have a base for closing our energy leaks!

haukepetersen commented:
on it.

haukepetersen commented on Makefile.features:
```diff
@@ -0,0 +1,8 @@
+DEFAULT_FEATURES += periph_pm
+
+# add available default features to required list
+FEATURES_REQUIRED += $(filter-out $(DISABLE_FEATURES), $(filter $(FEATURES_PROVIDED), $(DEFAULT_FEATURES)))
```
Wait, I think there is a conceptual error here: 
`FEATURES_REQUIRED != (FEATURES_PROVIDED - DISABLE_FEATURES)`
`FEATURES_REQUIRED` should filled by the dependency definitions in the `Makefile.dep`s, e.g. the `at86rf2xx` requires `periph_gpio` and `periph_spi`.

So I guess this should be something like `FEATURES_REAL` := `FEATURES_REQUIRED` + `DEFAULT_FEATURES` - `DISABLE_FEATURES`?!

kaspar030 replied:
```FEATURES_REQUIRED``` is being used to set the defines in riot_build.h.
```DEFAULT_FEATURES``` is used to always include a feature, *if available* (e.g., always require periph_pm)
```DISABLE_FEATURES``` can be used to override ```DEFAULT_FEATURES```(e.g., this time, build without periph/pm)

The file adds all features from "default features" that are available, unless they are in "disabled features", to the required list.

So in set operations, ``` FEATURES_REQUIRED += (DEFAULT_FEATURES | FEATURES_PROVIDED) - DISABLED_FEATURES ```. I think the last line in the file does that.

haukepetersen replied:
I think in your last statement we have the problem: `FEATURES_REQUIRED` should be `(DEFAULT_FEATURES & FEATURES_PROVIDED) - DISABLED_FEATURES`?!

kaspar030 replied:
hehe I think we need to define our operators. ;)

The make line adds all features from the default features that are provided and not disabled to the required features.

If during dependency processing other features are already in the required list, they stay in. Also, disabled_features does only disable default features.

haukepetersen replied:
in that case everything is just fine. Good thing we talked about it :-)

haukepetersen commented on boards/avsextrem/drivers/avsextrem-smb380.c:
```diff
@@ -1021,7 +1016,7 @@ void SMB380_enableNewDataInt(void)
      * interrupts
      */
-    SETBIT(lpm_prevent_sleep, LPM_PREVENT_SLEEP_ACCSENSOR);
+//    SETBIT(lpm_prevent_sleep, LPM_PREVENT_SLEEP_ACCSENSOR);
```
just remove these lines.

kaspar030 replied:
did

haukepetersen commented on boards/native/Makefile.features:
```diff
@@ -1,6 +1,7 @@
 # Put defined MCU peripherals here (in alphabetical order)
 FEATURES_PROVIDED += periph_cpuid
 FEATURES_PROVIDED += periph_hwrng
+FEATURES_PROVIDED += periph_pm
```
I would suggest moving this line to `cpu/native/Makefile.features`, as PM has not board specific configuration

kaspar030 replied:
did.

haukepetersen commented on core/include/reboot.h:
```diff
@@ -28,7 +30,10 @@
+static inline void reboot(void)
+{
+    pm_reboot();
+}
```
question: why don't we just throw out this file completely and use `pm_reboot` directly in `core_panic`?

haukepetersen replied:
Just seen below, `core_panic` is already using `pm_reboot` directly...

kaspar030 replied:
Yes. I kept the interface for now in order to keep the PR small, and to maybe, have the option to hook in things, like a message, at a central spot.

haukepetersen replied:
Hm: I would like to say it with your words: don't leave stuff there because we might need it again at some point in the future, reintroduce once really needed :-)

haukepetersen commented on core/kernel_init.c:
```diff
@@ -66,14 +68,9 @@ static void *idle_thread(void *arg)
-        }
+#ifdef FEATURE_PERIPH_PM
+        pm_set_lowest();
+#endif
```
in case PM is not present, wouldn't 
```c
#else
    while (1) {}
#endif
```
be more precise, so we prevent the scheduler from looping?

kaspar030 replied:
The scheduler is supposed to loop if there's no low-power mode.

haukepetersen replied:
I see, fine with me.

haukepetersen commented on core/panic.c:
```diff
@@ -75,7 +76,10 @@ NORETURN void core_panic(core_panic_t crash_code, const char *message)
+    pm_reboot();
+#else
+    /* DEVELHELP set => power off system */
+    pm_off();
```
can wen maybe get around the `weak` thing by handling this equal to the idle thread: in case the PM feature is provided we use it, if not simply go into a `while (1) {}` loop?! -> as PM present is the usual case anyway?

kaspar030 replied:
pm_off() has a weak default doing while(1) {}... Saves an ifdef here.

haukepetersen replied:
saw this later in the code, too -> ignore my comment above.

haukepetersen commented on cpu/k60/Makefile.features:
```diff
@@ -0,0 +1 @@
+FEATURES_PROVIDED += periph_pm
```
I's say it makes sense to move this to `cpu/kinetis_common/Makefile.features`, as all the kinetis CPUs share the PM stuff...

kaspar030 replied:
Well, currently they don't, and I don't have a board to test.

haukepetersen replied:
fine, will adapt this in a follow up PR.

haukepetersen commented on cpu/k60/include/periph_cpu.h:
```diff
@@ -0,0 +1,6 @@
+
+#define PM_NUM_MODES    (1U)
+
+#endif /* PERIPH_CPU_H */
```
again, move to kinetis common.

haukepetersen commented on cpu/nrf51/periph/pm.c:
```diff
@@ -0,0 +1,30 @@
+{
+    NRF_POWER->RAMON = 0;
+    NRF_POWER->SYSTEMOFF = 1;
+}
```
should be moved to `nrf5x_common`?!

kaspar030 replied:
One turns off the ram, one does not. Which is correct, and do they work on both?

haukepetersen replied:
Same here: will take care of this in a follow up PR.

haukepetersen commented on cpu/stm32_common/include/periph_cpu_common.h:
```diff
@@ -196,5 +196,9 @@ void gpio_init_analog(gpio_t pin);
 }
 #endif
 
+#if defined(CPU_FAM_STM32F1) || defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F2)
```
STM32F2 twice here?!

haukepetersen commented on cpu/x86/include/periph_cpu.h:
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2016 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This library is free software; you can redistribute it and/or
```
did you use a different license header here on purpose?

kaspar030 replied:
Nope, probably copied it from the next header I could find.

haukepetersen commented on drivers/include/periph/pm.h:
```diff
@@ -0,0 +1,143 @@
+ * @brief   Switches the MCU to a new power mode
+ *
+ * This function will be called by the idle thread after determining the lowest
+ * non-blocked mode.
```
Doc is off?! `kernel_init.c` simply calls `pm_set_lowest()`, does it not?

kaspar030 replied:
True, fixed.

kaspar030 commented:
> reintroduce once really needed :-)

Currently it is really needed to prevent this PR from going out of hand.

kaspar030 commented:
I just realized that in the current pm_set_lowest() impls, I don't deal with the "implicit idle mode". Will fix.

haukepetersen commented:
> Currently it is really needed to prevent this PR from going out of hand.

I see, so leave it as is and we can discuss in follow up PRs. So as I see it, all my comments are addressed (once you push your changes), anything else that is holding this PR back?

haukepetersen commented:
I am happy, please squash (up to you if you want to Murdock before hand...)

kaspar030 commented:
- squashed

kaspar030 commented:
- fixed pm mode order of saml21 and stm32 to comply with low mode -> low power.

haukepetersen commented:
re-ack, please squash

kaspar030 commented:
Looking at the build queue, I'll wait for this to be CIed and then squash if all but static tests are fine... I know there are unrelated doxygen warnings...

haukepetersen commented:
sounds like a plan

pwillem commented on drivers/include/periph/pm.h:
```diff
@@ -0,0 +1,143 @@
+ *   PM_NUM_MODES-1, the highest)
+ * - there is an implicit extra idle mode (which has the number PM_NUM_MODES)
+ * - individual power modes can be blocked/unblocked, e.g., by peripherals
+ * - if a mode is blocked, so are implicitly all lower modes
```
It seems that the implementation doesn't follow this rule. It selects the first unblocked power mode starting from the lowest mode [drivers/periph_common/pm.c](https://github.com/kaspar030/RIOT/blob/bb2fb6d28c3efda8db3d3582233be06034106f2c/drivers/periph_common/pm.c#L31)

``` c
void pm_set_lowest(void)
{
    pm_blocker_t blocker = (pm_blocker_t) pm_blocker;
    unsigned mode = 0;
    while (mode < PM_NUM_MODES) {
        if (! (blocker.val_u8[mode])) {
            break;
        }
        mode++;
    }
   ...
}
```

kaspar030 replied:
Ups. :)

Would you mind taking another look at the hopefully fixed logic?

pwillem replied:
If `blocker.val_u32=0` the new loop will try to set UINTN_MAX as mode
If you change 
```c
    while (mode--) {
```
to 
```c
    while (mode) {
        mode--;
```
The correct mode (0) will be selected

kaspar030 replied:
It is now

```
    while (mode) {
        if (blocker.val_u8[mode-1]) {
            break;
        }
        mode--;
    }
```

kaspar030 commented:
Talking offline with @haukepetersen, we decided that all the ```#ifdef FEATURE_PERIPH_PM``` was a little hard to follow, so I moved the layered PM code into a submodule.

Now the structure is as follows:

1. all MCU's need to have ```pm_off()``` ```and pm_reboot()```, although there are weak defaults
2. if an MCU provides the periph_pm feature, the idle thread will call ```pm_set_lowest()```. That's the hook of the kernel into power mode management. See e.g., native, as an example.
3. if the MCU is using a layered power mode approach, it can use the pm_layered module, which offers shared base functionality for blocking/unblocking low power modes. In this case, the MCU needs to implement pm_set(mode), and set PM_NUM_MODES in it's periph_cpu.conf. See e.g., samd21 for an example.

lebrush commented on sys/pm_layered/pm.c:
```diff
@@ -0,0 +1,95 @@
+    unsigned state = irq_disable();
+    if (blocker.val_u32 == pm_blocker.val_u32) {
+        DEBUG("pm: setting mode %u\n", mode);
+        pm_set(mode);
```
Power mode is set while interrupts are disabled. Does this mean that some MCU won't be able to wake up from interrupt when going to LP?

haukepetersen replied:
AFAIK waking up is independent from the global interrupt mask bit value on the platforms I know, so this should be fine here. But if we find a platform that has a different opinion about this, we need to change this part I guess...

lebrush replied:
Ok :-) Thanks!

lebrush commented on sys/pm_layered/pm.c:
```diff
@@ -0,0 +1,95 @@
+
+    /* set lowest mode if blocker is still the same */
+    unsigned state = irq_disable();
+    if (blocker.val_u32 == pm_blocker.val_u32) {
```
Why doing it twice? There are only 4 power modes, everything could be packed into the `irq_disable` part, right?

haukepetersen replied:
not sure I get the question, what is done twice here?

lebrush replied:
TL;DR: I missed the previous comment about the pm_layered, sorry. Now it makes sense. Ignore my comment. Thanks!

What I meant is: first you set `blocker` as `pm_blocker`, compute into which mode to go and then when the irq is disabled, it is verified that `blocker` and `pm_blocker` are still the same (safety check). Why not doing everything into the irq_disable if at most 4 iterations will occur? I assume: keep things within irq_disable blocks short :-)

In any case, I asked because the only reason I see to have 4 power modes is because of the `val_u32`, which is used: to "atomically" check if the blocked modes changed within the call of `pm_set_lowest` and to set all block modes to 0 in the power_off method (in the latest we are not in a hurry, one could set all to 0 in a loop as well).

lebrush commented on sys/pm_layered/pm.c:
```diff
@@ -0,0 +1,95 @@
+
+void pm_block(unsigned mode)
+{
+    assert(pm_blocker.val_u8[mode] != 255);
```
if it safe to do it as an assert or should an `if` be required:

```C
unsigned state = irq_disable();
if (pm_blocker.val_u8[mode] < 255) {
    pm_blocker.val_u8[mode]++;
}
irq_restore(state);
```

haukepetersen replied:
I would say the `assert` is the way to go here. If we just don't increment the value but continue as if nothing was wrong, we get an inconsistent state of the blocker value, and that is not good... Also it is from todays view very unlikely, that this assert will ever be triggered...

lebrush replied:
Agree

haukepetersen commented:
> I've a question to raise: the topic of "what if more than 4 power modes are needed" got dropped

This is now handled implicitly by using (or not using) the `pm_layered` module: if one wants to use the `pm_blocker_t` and everything that comes with it for implementing PM on a target CPU, one is bound to max 4 power levels. If one wants to do anything else, just don't use the `pm_layered` module and the implementation can do whatever seems best for the CPU, all that MUST be done is implementing `pm_set_lowest`.

haukepetersen commented:
Oh, maybe for transparency: @kaspar030 and me had some offline calls over the last day, so all my concerns/opinions stated throughout this discussion here are addressed...

haukepetersen commented:
@kaspar030: something with the `FEATURE_PROVIDED` var seems to go wrong, Murdock tried to build applications for boards that don't have specific features...

kaspar030 commented:
> @kaspar030: something with the FEATURE_PROVIDED var seems to go wrong, Murdock tried to build applications for boards that don't have specific features...

Shitto, we need to redo that logic. ;)

Anyhow, I squashed in a change that fixes it locally. I had moved the features processing up before the include of Makefile.buildtest, that seems to break it.

lebrush commented:
Might it be related to the issue in Murdock with #5573? (it's compiling for platforms which have no i2c even if it's required)

kaspar030 commented:
All green! @haukepetersen would you do me the honour of pressing the button? At least half of the concept has been devised by you!

haukepetersen commented:
Oh yeah -> let's power down now :-)

miri64 commented on drivers/periph_common/pm.c:
```diff
@@ -0,0 +1,33 @@
+
+void __attribute__((weak)) pm_set_lowest(void) {}
+
+void __attribute__((weak)) pm_off(void)
```
I know this already got merged, but doc states `pm_off()` *must* be provided by every platform. So why is this here?

haukepetersen replied:
Will be taken care of in a follow up PR...

OlegHahm commented:
I don't have any concrete objections against the content of this PR, but I think that this kind of crucial feature should have required more than one ACK.

immesys commented:
@kaspar030 sorry for being dense, but where should I implement the pm_set_lowest for my platform? Is this something that goes in boards/X/? My platform is SAMR21 based so would I put it in the cpu/samd21/periph/pm.c? Is there documentation somewhere I am missing?
